### ISSUE TITLE
[BUG] Segmentation Fault with multiple tests with AudioBundle and RenderBundle

### ISSUE THREAD

**Issue by azriel91** [amethyst/amethyst#1595]
*see also: https://github.com/retep998/winapi-rs/issues/770*

## Description

Given:

* There are at least two `#[test]`s that use `amethyst_test::AmethystApplication`
* Both are in the same binary (i.e. tests are in same module)
* Both use `AudioBundle::default()` and `RenderBundle`
* Run this on Windows

Then we get a segfault:

```rust
$ cargo test -p amethyst_test audio
   Compiling amethyst_test v0.1.0 (D:\work\github\amethyst\amethyst\amethyst_test)
    Finished dev [unoptimized + debuginfo] target(s) in 28.10s
     Running target\debug\deps\amethyst_test-d4e93333e68ef2fd.exe

running 2 tests
test amethyst_application::test::audio_two ... ok
error: test failed, to rerun pass '-p amethyst_test --lib'
Segmentation fault
```

## Reproduction Steps

```bash
git remote add azriel91 git@github.com:azriel91/amethyst.git
git fetch azriel91 bug/1595/segfault-with-multiple-audio-tests
git checkout bug/1595/segfault-with-multiple-audio-tests
cargo test -p amethyst_test audio
```

## What You Expected to Happen

No segfault.

# Logs

```
     Running target\debug\deps\amethyst_test-d4e93333e68ef2fd.exe

running 2 tests
[INFO][winit::platform::platform::window] Guessed window DPI factor: 1
[INFO][amethyst::app] Initializing Amethyst...
[INFO][amethyst::app] Version: 0.10.0
[INFO][amethyst::app] Platform: x86_64-pc-windows-msvc
[INFO][amethyst::app] Amethyst git commit: 7771ffd018256163fb13fec585668ac8844dba3c
[INFO][amethyst::app] Rustc version: 1.36.0-nightly Nightly
[INFO][amethyst::app] Rustc git commit: e305df1846a6d985315917ae0c81b74af8b4e641
[INFO][amethyst::app] Engine is shutting down
test amethyst_application::test::audio_two ... ok
[INFO][winit::platform::platform::window] Guessed window DPI factor: 1
[INFO][amethyst::app] Initializing Amethyst...
[INFO][amethyst::app] Version: 0.10.0
[INFO][amethyst::app] Platform: x86_64-pc-windows-msvc
[INFO][amethyst::app] Amethyst git commit: 7771ffd018256163fb13fec585668ac8844dba3c
[INFO][amethyst::app] Rustc version: 1.36.0-nightly Nightly
[INFO][amethyst::app] Rustc git commit: e305df1846a6d985315917ae0c81b74af8b4e641
error: test failed, to rerun pass '-p amethyst_test --lib'
PS D:\work\github\amethyst\amethyst> $LastExitCode
-1073741819
```

## Your Environment

* Windows 10 x64
* Rust 1.36 `nightly` (but also seen on Rust 1.34.1 `stable`)
* Amethyst master: `7771ffd0`

## Additional Context

**If you run either one of those tests, there will NOT be a segfault.** It has to be two or more.

Also, the exit code `-1073741819` = unsigned 32-bit `0xC0000005` = `STATUS_ACCESS_VIOLATION`

* https://github.com/commercialhaskell/stackage/issues/3385#issuecomment-373714206
* https://stackoverflow.com/a/53485066/1576773

    > The return code of 0xC0000005 means `STATUS_ACCESS_VIOLATION`. (You can find this and other NT status codes on MSDN: [NTSTATUS Values](https://msdn.microsoft.com/en-us/library/cc704588.aspx).) The error happens because `NULL` is outside the range of valid addresses for your program. Before dereferencing a pointer variable, you must assign it the address of a valid object.

* https://github.com/rust-lang/rust/issues/33434 Seen in Rust
* https://github.com/rust-lang/rust/issues/50604 64 bit value is: `3221225477`

**Comment by azriel91** [comment]
Okay, we have a stack trace!

**The short:**

`cpal::Device#clone()` segfaults.

* crate: `cpal-0.8.2`
* file: `src/wasapi/device.rs:545`

**The long:**

1. Fails when cloning `amethyst_audio::output::Output`:

    https://github.com/amethyst/amethyst/blob/209f8986c3a2e033c6a177da345d18daeca8d59e/amethyst_audio/src/systems/audio.rs#L129-L132

2. This calls down to `cpal::Device#clone()`, which does this `unsafe` thing.

    https://github.com/tomaka/cpal/blob/0354548426f7b4c48392cae9876a3e77ea5c1c40/src/wasapi/device.rs#L542-L547

    ```rust
    impl Clone for Device {
    #[inline]
    fn clone(&self) -> Device {
        unsafe {
            (*self.device).AddRef();
        }
    ```

    The `AddRef()` encompasses where the segfault happens, let's go down one more.

    This is: `winapi-0.3.7`, file `src/macros.rs`:

    https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/macros.rs#L199-L203

    ```rust
    (@method fn $method:ident($($p:ident : $t:ty,)*) -> $rtr:ty) => (
        #[inline] pub unsafe fn $method(&self, $($p: $t,)*) -> $rtr {
            ((*self.lpVtbl).$method)(self as *const _ as *mut _, $($p,)*)
        }
    );
    ```

3. Uh okay, how did it suddenly jump from `AddRef` to that macro?

    `(*self.device)` in 2, [is a `winapi::um::mmdeviceapi::IMMDevice`](https://github.com/tomaka/cpal/blob/0354548426f7b4c48392cae9876a3e77ea5c1c40/src/wasapi/device.rs#L76). Which is [this](https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/um/mmdeviceapi.rs#L121) (*whoa! is that even Rust?!*).

    ```rust
    RIDL!{#[uuid(0xd666063f, 0x1587, 0x4e43, 0x81, 0xf1, 0xb9, 0x48, 0xe8, 0x07, 0x36, 0x3f)]
    interface IMMDevice(IMMDeviceVtbl): IUnknown(IUnknownVtbl) {
        // ..
    }
    ```

    Anyway, [it `Deref`s to `IMMDeviceVtbl`](https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/macros.rs#L190-L198), which, as far as I can tell... uh, I can't tell very far -- because a textual search of the repository only shows that usage, and no declaration / definition.

    Oh wait. it's because that outer macro defines *both the outer interface and the inner thing.* ([1](https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/macros.rs#L174-L179), [2](https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/macros.rs#L211-L222)).

    Okay. Before we get too carried away, we're trying to figure out how we got to the `Clone`.

    If we just text search for "Clone" in that `winapi` `macros.rs` file, there are 6 results, first 4 being for "union", and the [last two for "struct"](https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/macros.rs#L377-L380).

    ```rust
    impl Clone for $name {
        #[inline]
        fn clone(&self) -> $name { *self }
    }
    ```

4. So. We need to figure out what thing is being cloned that shouldn't be.

    That is either one of [these things](https://github.com/retep998/winapi-rs/blob/0.3/src/um/mmdeviceapi.rs#L121-L138):

    ```rust
    fn Activate(
        iid: REFIID,
        dwClsCtx: DWORD,
        pActivationParams: *mut PROPVARIANT,
        ppInterface: *mut LPVOID,
    ) -> HRESULT,
    fn OpenPropertyStore(
        stgmAccess: DWORD,
        ppProperties: *mut *mut IPropertyStore,
    ) -> HRESULT,
    fn GetId(
        ppstrId: *mut LPWSTR,
    ) -> HRESULT,
    fn GetState(
        pdwState: *mut DWORD,
    ) -> HRESULT,
    ```

    or, something in the [`IUnknown` or `IUnknownVtbl`](https://github.com/retep998/winapi-rs/blob/c45cef77c93ef3a0f8921a300dc58753fc463297/src/um/unknwnbase.rs#L11).

The plot thickens.

<details>

<summary>Stack trace</summary>

```
winapi::um::unknwnbase::IUnknown::AddRef() Line 201
    at C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.3.7\src\macros.rs(201)
cpal::cpal_impl::device::{{impl}}::clone(cpal::cpal_impl::device::Device * self) Line 545
    at C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.8.2\src\wasapi\device.rs(545)
cpal::{{impl}}::clone(cpal::Device * self) Line 161
    at C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.8.2\src\lib.rs(161)
amethyst_audio::output::{{impl}}::clone(amethyst_audio::output::Output * self) Line 22
    at D:\work\github\amethyst\amethyst\amethyst_audio\src\output.rs(22)
amethyst_audio::systems::audio::{{impl}}::setup(amethyst_audio::systems::audio::AudioSystem * self, shred::res::Resources * res) Line 131
    at D:\work\github\amethyst\amethyst\amethyst_audio\src\systems\audio.rs(131)
shred::system::{{impl}}::setup<amethyst_audio::systems::audio::AudioSystem>(amethyst_audio::systems::audio::AudioSystem * self, shred::res::Resources * res) Line 138
    at C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\shred-0.7.2\src\system.rs(138)
shred::dispatch::stage::Stage::setup(shred::res::Resources * self) Line 85
    at C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\shred-0.7.2\src\dispatch\stage.rs(85)
shred::dispatch::dispatcher::Dispatcher::setup(shred::res::Resources * self) Line 21
    at C:\Users\azrie\.cargo\registry\src\github.com-1ecc6299db9ec823\shred-0.7.2\src\dispatch\dispatcher.rs(21)
amethyst::game_data::{{impl}}::build(amethyst::game_data::GameDataBuilder<f32> self, specs::world::World * world) Line 277
    at D:\work\github\amethyst\amethyst\src\game_data.rs(277)
amethyst::app::ApplicationBuilder<amethyst_test::state::sequencer_state::SequencerState<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>>, amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>, amethyst::state_event::StateEventReader<alloc::string::String>>::build<amethyst_test::state::sequencer_state::SequencerState<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>>,amethyst::game_data::GameData,amethyst::state_event::StateEvent<alloc::string::String>,amethyst::state_event::StateEventReader<alloc::string::String>,amethyst::game_data::GameDataBuilder<f32>>(amethyst::app::ApplicationBuilder<amethyst_test::state::sequencer_state::SequencerState<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>>, amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>, amethyst::state_event::StateEventReader<alloc::string::String>> self, amethyst::game_data::GameDataBuilder<f32> init) Line 870
    at D:\work\github\amethyst\amethyst\src\app.rs(870)
amethyst_test::amethyst_application::AmethystApplication<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>, amethyst::state_event::StateEventReader<alloc::string::String>>::build_application<amethyst::state_event::StateEvent<alloc::string::String>,amethyst::state_event::StateEventReader<alloc::string::String>,amethyst_test::state::sequencer_state::SequencerState<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>>>(amethyst_test::state::sequencer_state::SequencerState<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>> first_state, amethyst::game_data::GameDataBuilder<f32> game_data, alloc::vec::Vec<alloc::boxed::Box<FnMut<(mut specs::world::World*)>>> resource_add_fns) Line 259
    at D:\work\github\amethyst\amethyst\amethyst_test\src\amethyst_application.rs(259)
amethyst_test::amethyst_application::AmethystApplication<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>, amethyst::state_event::StateEventReader<alloc::string::String>>::build_internal<amethyst::state_event::StateEvent<alloc::string::String>,amethyst::state_event::StateEventReader<alloc::string::String>>((alloc::vec::Vec<boxfnonce::send::SendBoxFnOnce<(amethyst::game_data::GameDataBuilder<f32>), core::result::Result<amethyst::game_data::GameDataBuilder<f32>, amethyst_error::Error>>>, alloc::vec::Vec<alloc::boxed::Box<FnMut<(mut specs::world::World*)>>>, alloc::vec::Vec<boxfnonce::send::SendBoxFnOnce<(), alloc::boxed::Box<State<amethyst::game_data::GameData, amethyst::state_event::StateEvent<alloc::string::String>>>>>)) Line 239
    at D:\work\github\amethyst\amethyst\amethyst_test\src\amethyst_application.rs(239)
amethyst_test::amethyst_application::{{impl}}::run::{{closure}}<amethyst::state_event::StateEvent<alloc::string::String>,amethyst::state_event::StateEventReader<alloc::string::String>>(closure) Line 298
    at D:\work\github\amethyst\amethyst\amethyst_test\src\amethyst_application.rs(298)
std::sys_common::backtrace::__rust_begin_short_backtrace<closure,core::result::Result<(), amethyst_error::Error>>(closure f) Line 136
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\sys_common\backtrace.rs(136)
std::thread::{{impl}}::spawn_unchecked::{{closure}}::{{closure}}<closure,core::result::Result<(), amethyst_error::Error>>(closure) Line 469
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\thread\mod.rs(469)
std::panic::{{impl}}::call_once<core::result::Result<(), amethyst_error::Error>,closure>(std::panic::AssertUnwindSafe<closure> self) Line 309
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\panic.rs(309)
std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure>,core::result::Result<(), amethyst_error::Error>>(unsigned char * data) Line 293
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\panicking.rs(293)
panic_unwind::__rust_maybe_catch_panic() Line 85
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\/src\libpanic_unwind\lib.rs(85)
std::panicking::try<core::result::Result<(), amethyst_error::Error>,std::panic::AssertUnwindSafe<closure>>(std::panic::AssertUnwindSafe<closure> f) Line 272
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\panicking.rs(272)
std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,core::result::Result<(), amethyst_error::Error>>(std::panic::AssertUnwindSafe<closure> f) Line 390
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\panic.rs(390)
std::thread::{{impl}}::spawn_unchecked::{{closure}}<closure,core::result::Result<(), amethyst_error::Error>>(closure) Line 468
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libstd\thread\mod.rs(468)
core::ops::function::FnOnce::call_once<closure,()>(closure *) Line 231
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\libcore\ops\function.rs(231)
alloc::boxed::{{impl}}::call_once<(),FnOnce<()>>() Line 704
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\liballoc\boxed.rs(704)
alloc::boxed::{{impl}}::call_once() Line 704
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\src\liballoc\boxed.rs(704)
std::sys_common::thread::start_thread() Line 13
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\/src\libstd\sys_common\thread.rs(13)
std::sys::windows::thread::{{impl}}::new::thread_start() Line 47
    at /rustc/e305df1846a6d985315917ae0c81b74af8b4e641\/src\libstd\sys\windows\thread.rs(47)
```

</details>

**Comment by retep998** [comment]
Lots of people rely on the COM interfaces in `winapi` without issue, so I doubt there is anything wrong with how `AddRef` is implemented. Also you may be confused about how things are working, because calling `AddRef` does not clone anything, it merely calls the `AddRef` function from the vtbl (which generally just increments a reference count).

One possible explanation is a threading issue. You might be sending COM objects to a thread that isn't initialized for COM or you might be sending COM objects across threads that were created in an apartment thread and therefore are not thread safe.
Another possible explanation is that you're not calling `AddRef` or `Release` at the right times and the COM object is destroyed while you're still using it.

**Comment by azriel91** [comment]
Thank you :bow:! That pinpointed the issue. Let's break it down:

**Background**

1. On Windows, a thread must be initialized to use COM objects before it uses COM objects.
2. `cpal::default_output_device()` is one such usage, which internally [COM initializes the thread](https://github.com/tomaka/cpal/blob/0354548426f7b4c48392cae9876a3e77ea5c1c40/src/wasapi/device.rs#L609).
3. `amethyst::audio::Output` owns a `Device`, obtained from `cpal::default_output_device()`.
4. This is first obtained through `AudioBundle::default()`, then passed to `AudioSystem::new(output)`
5. In `AudioSystem#setup(..)` this clones `Output` and adds it to the `Resources` -- internally this accesses the `IMMDevice` COM object and increments its reference counter.

**Threads**

1. Amethyst has its main thread. The main thread is the one that invokes `Dispatcher#setup`, which invokes `System#setup` on the various systems.
2. Amethyst also has a pool of worker threads, each is free to pick up a `System#run`.
3. `amethyst_test` wraps the Amethyst application in a thread to avoid a bug in `mesa` that caused segfaults on operating systems that use X (e.g. Ubuntu) (https://github.com/rust-windowing/glutin/issues/1034).
4. `cargo test` wraps each test in a thread.

The 1. and 3. above are the same thread.

---

The Problem:

* In the originally reported issue, code that initializes the `AudioBundle` is in a `#[test]` function, meaning the thread in `Threads#4` is COM initialized, but not `Threads#1` (or 3.).
* This means when we run the application, the `AudioSystem#setup` is gonna hit `Background#5`, which causes the `STATUS_ACCESS_VIOLATION`.
* If we use `.with_bundle_fn(AudioBundle::default())`, it still does not COM initialize `Threads#1`, because the `Dispatcher#setup` is done in `Threads#4` before being pushed to `#Threads#1`.

More:

* If we hackily COM initialize `Thread#1`, by instantiating *a separate instance of* `AudioBundle::default()` in that thread, **the segfault moves** from `System#setup` to where the Amethyst application is dropped.

    This is because COM initialization is not the only thing that a thread needs to have, but it must also have proper permissions to access the COM object -- and our cloning of the `amethyst::audio::Output` is invalidly sharing the `cpal::Device` around.

    For interest, it segfaults at [`cpal::Device#drop`](https://github.com/tomaka/cpal/blob/0354548426f7b4c48392cae9876a3e77ea5c1c40/src/wasapi/device.rs#L572).


* The segfault happened on `System#setup`, but we are still (invalidly ?) accessing the underlying `cpal::Device`, as `amethyst::audio::Output` is used by `AudioSystem`, which is run by any of the worker threads in the pool -- we have not guaranteed that these threads are COM initialized.

---

To prevent this error from happening (with good practice), we should do one of:

* Keep `AudioSystem` as a thread local system, and make sure that system is the only one instantiating and accessing the `cpal::Device`.
* Ensure the current thread is COM initialized in: `AudioSystem#setup` **and** `AudioSystem#run`, **AND** don't simply call `clone()` on `amethyst::audio::Output`.

    - Is the `Clone` implementation incorrect?
    - I'm not too sure how you're meant to share a `cpal::Device` across threads.

**Comment by Xaeroxe** [comment]
Ultimately this does amount to a bug in upstream dependencies, so while we can implement our own work around, ultimately we should be seeking to drive API changes upstream.

**Comment by azriel91** [comment]
> * Sorry, this _should_ COM initialize `Threads#1`, `AmethystApplication#build_internal` is called inside the spawned thread. It still segfaulted, this time when [accessing the `IMMDeviceEnumerator`](https://github.com/tomaka/cpal/blob/0354548426f7b4c48392cae9876a3e77ea5c1c40/src/wasapi/device.rs#L722).
>   That's when creating `AudioBundle::default()` in `Threads#1`. Hm, this one I didn't expect.

I think this is because `ENUMERATOR.0` is created in `lazy_static`, in `Threads#4`, then accessed in `Threads#1`

**Comment by jojolepro** [comment]
Wow nice work debugging all of this!

**Comment by azriel91** [comment]
> `ENUMERATOR.0` is created in `lazy_static`, in `Threads#4`, then accessed in `Threads#1`

in Windows COM objects, given a program has 4 threads: thread 0 (main), thread 1, 2, 3

if thread 1:

1. is COM initialized as `COINIT_MULTITHREADED`
2. creates a windows COM object in `lazy_static!`

what makes it illegal for thread 2 (which is also COM initialized), to access that COM object through the `lazy_static` reference?
(program segfaults: `STATUS_ACCESS_VIOLATION`)

**Comment by azriel91** [comment]
Tried:

* Commenting out `Enumerator#drop`. `parent` methods were still `???` on second test.
* `(*enumerator).AddRef();` in `ENUMERATOR` constructor. `parent` methods were still `???` on second test.
* Breakpoint on `Enumerator#drop`. This was never hit.
* Breakpoint on `Enumerator#drop` with `#[inline]` removed. This was never hit.
* Breakpoint on `winapi::um::unknwnbase::IUnknown::Release`. This was [hit by `cpal::Device#drop`](https://github.com/tomaka/cpal/blob/0354548426f7b4c48392cae9876a3e77ea5c1c40/src/wasapi/device.rs#L574) twice during the first test, the the second test runs and hits the `STATUS_ACCESS_VIOLATION`.

...and tried removing the sub-thread wrapper around the `AmethystApplication`, and use canonical `cpal` from crates.io.

It worked 😯

Theory, because `AmethystApplication` is run in its own thread, windows initializes the COM object in some memory location that gets cleaned up when that thread is `joined`, but the `ENUMERATOR` static ref still points to that memory location. `Enumerator#drop` is never called, because `ENUMERATOR` lives in application memory which is still valid.

**Comment by azriel91** [comment]
Implemented option 1 in #1599.
